// @storybook/angular/dist/server/framework-preset-angular.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.webpack = webpack;

var _path = _interopRequireDefault(require("path"));

var _webpack = require("webpack");

var _ts_config = _interopRequireDefault(require("./ts_config"));

var _createForkTsCheckerPlugin = _interopRequireDefault(require("./create-fork-ts-checker-plugin"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function webpack(config, {
  configDir
}) {
  const tsLoaderOptions = (0, _ts_config.default)(configDir);
  return _objectSpread({}, config, {
    module: _objectSpread({}, config.module, {
      rules: [...config.module.rules, {
        test: /\.tsx?$/,
        use: [{
          loader: require.resolve('ts-loader'),
          options: tsLoaderOptions
        }, require.resolve('angular2-template-loader')]
      }, {
        test: /[/\\]@angular[/\\]core[/\\].+\.js$/,
        parser: {
          system: true
        }
      }, {
        test: /\.html$/,
        loader: 'raw-loader',
        exclude: /\.async\.html$/
      }, {
        test: /\.s(c|a)ss$/,
        use: [require.resolve('raw-loader'), require.resolve('sass-loader')]
      }]
    }),
    resolve: _objectSpread({}, config.resolve, {
      extensions: ['.ts', '.tsx', ...config.resolve.extensions]
    }),
    plugins: [...config.plugins, // See https://github.com/angular/angular/issues/11580#issuecomment-401127742
    new _webpack.ContextReplacementPlugin(/@angular(\\|\/)core(\\|\/)(fesm5|bundles)/, _path.default.resolve(__dirname, '..')), (0, _createForkTsCheckerPlugin.default)(tsLoaderOptions)]
  });
}
